- Select dataset
- Perform data preparation and cleaning using Pandas & Numpy
- Perform exploratory analysis & visualization using Matplotlib and Seaborn
- Ask and Answer questions about the data in a Jupyter Notebook
- Summarize your inference and write a conclusion
- Document, publish and present your Jupyter Notebook Online
- Create interactive Dashboard 

# American Values Data Analysis

TODO- talk about EDA

TODO- talk about the dataset (source, what it contains, how it will be useful)

- Source: Five Thirty Eight
- Can be useful to understand American values how 

# Data Preparation and Cleaning

- Load the file using pandas
- Look at some information about the file about the data and columns
- Fix any missing or incorrect values


## Load the file using pandas

In [2]:
import numpy as np
import pandas as pd

df = pd.read_csv('nonvoters_data.csv')


df.head()

RespId  weight  Q1  Q2_1  Q2_2  Q2_3  Q2_4  Q2_5  Q2_6  Q2_7  ...  Q30  \
0  470001  0.7516   1     1     1     2     4     1     4     2  ...    2   
1  470002  1.0267   1     1     2     2     3     1     1     2  ...    3   
2  470003  1.0844   1     1     1     2     2     1     1     2  ...    2   
3  470007  0.6817   1     1     1     1     3     1     1     1  ...    2   
4  480008  0.9910   1     1     1    -1     1     1     1     1  ...    1   

   Q31  Q32  Q33  ppage                 educ   race  gender     income_cat  \
0  NaN  1.0  NaN     73              College  White  Female       $75-125k   
1  NaN  NaN  1.0     90              College  White  Female  $125k or more   
2  NaN  2.0  NaN     53              College  White    Male  $125k or more   
3  NaN  1.0  NaN     58         Some college  Black  Female        $40-75k   
4 -1.0  NaN  NaN     81  High school or less  White    Male        $40-75k   

   voter_category  
0          always  
1          always  
2        sporadic  
3        sporadic  
4          always  

[5 rows x 119 columns]

## Look at information about the data and columns

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5836 entries, 0 to 5835
Columns: 119 entries, RespId to voter_category
dtypes: float64(23), int64(91), object(5)
memory usage: 5.3+ MB


We have 119 columns in our dataset and 5,836 rows. 

There's too many columns to see the full number and so, to find out information about the columns, [we refer to the codebook for the dataset](https://github.com/fivethirtyeight/data/blob/master/non-voters/nonvoters_codebook.pdf).

This analysis will focus on questions two and three in the study. - we'll filter the dataframe to only include the pertinent columns.

This corresponds with 

Now we'll rename those columns and then create a new dataframe with the new column names and the socioeconomic columns.

2. In your view, how important are each of the following to being a good American?

   1. Voting in elections
   2. Serving on a jury
   3. Following what happens in government and politics
   4. Displaying the American flag
   5. Participating in the U.S. Census every ten years
   6. Knowing the Pledge of Allegiance
   7. Supporting the military
   8. Respecting the opinions of those who disagree with you
   9. Believing in God
   10. Protesting if you believe government actions are wrong
    
Responses
1. Very important
2. Somewhat important
3. Not so important
4. Not at all important


3. How much do you agree or disagree with the following statements?

   1. Systemic racism is a problem in the United States.
   2. Systemic racism in policing is a bigger problem than violence and vandalism
   in protests.
   1. Society as a whole has become too soft and feminine.
   2. The mainstream media is more interested in making money than telling the truth.
   3. Traditional parties and politicians don’t care about people like me.
   4. The way people talk needs to change with the times to be more sensitive to people with different backgrounds.

Responses
1. Strongly agree
2. Somewhat agree
3. Somewhat disagree
4. Strongly disagree

In [4]:
# Rename the columns

df = df.rename(columns={"Q2_1":"Voting",
        "Q2_2": "Jury",
        "Q2_3": "Govt_Politics",
        "Q2_4": "American_Flag",
        "Q2_5": "US_Census",
        "Q2_6": "Pledge",
        "Q2_7":"Military Support",
        "Q2_8": "Respecting_Others",
        "Q2_9":"Believing in God",
        "Q2_10":"Protesting", 
        "Q3_1": "Racism_US",
        "Q3_2": "Racism_Policing",
        "Q3_3":"Society_Whole",
        "Q3_4":"Mainstream_Media",
        "Q3_5":"Traditional_Parties",
        "Q3_5":"Way_People_Talk"})

df


RespId  weight  Q1  Voting  Jury  Govt_Politics  American_Flag  \
0     470001  0.7516   1       1     1              2              4   
1     470002  1.0267   1       1     2              2              3   
2     470003  1.0844   1       1     1              2              2   
3     470007  0.6817   1       1     1              1              3   
4     480008  0.9910   1       1     1             -1              1   
...      ...     ...  ..     ...   ...            ...            ...   
5831  477662  1.1916   1       1     3              1              3   
5832  477663  1.4623   1       1     1              1              2   
5833  488322  0.9252   1       1     2              1              3   
5834  488325  2.6311   1       2     2              2              2   
5835  477666  1.6218   1       1     3              2              3   

      US_Census  Pledge  Military Support  ...  Q30  Q31  Q32  Q33  ppage  \
0             1       4                 2  ...    2  NaN  1.0  NaN     73   
1             1       1                 2  ...    3  NaN  NaN  1.0     90   
2             1       1                 2  ...    2  NaN  2.0  NaN     53   
3             1       1                 1  ...    2  NaN  1.0  NaN     58   
4             1       1                 1  ...    1 -1.0  NaN  NaN     81   
...         ...     ...               ...  ...  ...  ...  ...  ...    ...   
5831          1       2                 2  ...    2  NaN  1.0  NaN     27   
5832          1       2                 1  ...    2  NaN  2.0  NaN     59   
5833          1       1                 2  ...    2  NaN  1.0  NaN     51   
5834          2       2                 2  ...    3  NaN  NaN  1.0     22   
5835          1       1                 2  ...    5  NaN  NaN  2.0     22   

                     educ         race  gender      income_cat  voter_category  
0                 College        White  Female        $75-125k          always  
1                 College        White  Female   $125k or more          always  
2                 College        White    Male   $125k or more        sporadic  
3            Some college        Black  Female         $40-75k        sporadic  
4     High school or less        White    Male         $40-75k          always  
...                   ...          ...     ...             ...             ...  
5831         Some college     Hispanic    Male         $40-75k          always  
5832  High school or less        White  Female   $125k or more    rarely/never  
5833              College  Other/Mixed    Male   $125k or more        sporadic  
5834  High school or less        Black  Female  Less than $40k          always  
5835  High school or less        Black  Female  Less than $40k          always  

[5836 rows x 119 columns]

In [5]:
# Create a copy of our dataframe 

df = df.copy().loc[:,["Voting",
                "Jury",
                "Govt_Politics",
                "American_Flag",
                "US_Census",
                "Pledge",
                "Military Support",
                "Respecting_Others",
                "Believing in God",
                "Protesting",
                "Racism_US",
                "Racism_Policing",
                "Society_Whole",
                "Mainstream_Media",
                "Way_People_Talk",
                "ppage",
                "educ",
                "race",
                "gender",
                "income_cat",
                "voter_category"]]

We'll get a list of columns to see the 21 columns that are in our new, copied dataset. 

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5836 entries, 0 to 5835
Data columns (total 21 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Voting             5836 non-null   int64 
 1   Jury               5836 non-null   int64 
 2   Govt_Politics      5836 non-null   int64 
 3   American_Flag      5836 non-null   int64 
 4   US_Census          5836 non-null   int64 
 5   Pledge             5836 non-null   int64 
 6   Military Support   5836 non-null   int64 
 7   Respecting_Others  5836 non-null   int64 
 8   Believing in God   5836 non-null   int64 
 9   Protesting         5836 non-null   int64 
 10  Racism_US          5836 non-null   int64 
 11  Racism_Policing    5836 non-null   int64 
 12  Society_Whole      5836 non-null   int64 
 13  Mainstream_Media   5836 non-null   int64 
 14  Way_People_Talk    5836 non-null   int64 
 15  ppage              5836 non-null   int64 
 16  educ               5836 non-null   object


We can see that we have fifteen questions for our dataset (numeric). 

The other six columns represent socioeconomic data points. Aside from age, these are all object data types (or categorical data).



In [11]:
df.describe()

Voting         Jury  Govt_Politics  American_Flag    US_Census  \
count  5836.000000  5836.000000    5836.000000    5836.000000  5836.000000   
mean      1.246402     1.705106       1.638280       2.175977     1.277245   
std       0.660253     0.866346       0.765741       1.091391     0.626386   
min      -1.000000    -1.000000      -1.000000      -1.000000    -1.000000   
25%       1.000000     1.000000       1.000000       1.000000     1.000000   
50%       1.000000     2.000000       2.000000       2.000000     1.000000   
75%       1.000000     2.000000       2.000000       3.000000     1.000000   
max       4.000000     4.000000       4.000000       4.000000     4.000000   

            Pledge  Military Support  Respecting_Others  Believing in God  \
count  5836.000000       5836.000000        5836.000000       5836.000000   
mean      1.805517          1.491604           1.464016          2.093557   
std       1.011524          0.808120           0.674857          1.245036   
min      -1.000000         -1.000000          -1.000000         -1.000000   
25%       1.000000          1.000000           1.000000          1.000000   
50%       1.000000          1.000000           1.000000          2.000000   
75%       2.000000          2.000000           2.000000          3.000000   
max       4.000000          4.000000           4.000000          4.000000   

        Protesting    Racism_US  Racism_Policing  Society_Whole  \
count  5836.000000  5836.000000      5836.000000    5836.000000   
mean      2.019363     1.904044         2.384510       2.630055   
std       0.946632     1.067898         1.193323       1.118382   
min      -1.000000    -1.000000        -1.000000      -1.000000   
25%       1.000000     1.000000         1.000000       2.000000   
50%       2.000000     2.000000         2.000000       3.000000   
75%       3.000000     3.000000         3.000000       4.000000   
max       4.000000     4.000000         4.000000       4.000000   

       Mainstream_Media  Way_People_Talk        ppage  
count       5836.000000      5836.000000  5836.000000  
mean           1.862406         2.015764    51.693797  
std            0.986499         0.868229    17.071561  
min           -1.000000        -1.000000    22.000000  
25%            1.000000         1.000000    36.000000  
50%            2.000000         2.000000    54.000000  
75%            2.000000         3.000000    65.000000  
max            4.000000         4.000000    94.000000

We see that that there were 5,836 responses in our dataset.  

For the Likert Scale questions, answers range from -1 to 4. It's unclear why -1 is one of the options and might be an error. 

We can also see that the age of our respondents ranges from 17 to 94 with an average age of about 52. 

## Fix any missing or incorrect values


In [13]:
# Are we missing any values?

df.isna().sum()

Voting               0
Jury                 0
Govt_Politics        0
American_Flag        0
US_Census            0
Pledge               0
Military Support     0
Respecting_Others    0
Believing in God     0
Protesting           0
Racism_US            0
Racism_Policing      0
Society_Whole        0
Mainstream_Media     0
Way_People_Talk      0
ppage                0
educ                 0
race                 0
gender               0
income_cat           0
voter_category       0
dtype: int64

In [28]:
# Create list of survey data columns

survey_data_columns = []
for column in df.columns:
    if df[column].dtypes == "int64" and column != 'ppage':
        survey_data_columns.append(column)

In [29]:
# TO DO Restructure data


df.apply(pd.value_counts)

# Filter for just response data

# Drop all values of -1

# Likert scale data (make agreeable)

# Create two lists for questions two and three, restructure data that way

# Age (bins)

# make the bins by quartile



Voting    Jury  Govt_Politics  American_Flag  US_Census  Pledge  \
-1              35.0    36.0           36.0           34.0       28.0    31.0   
1             4758.0  2811.0         2772.0         1966.0     4513.0  2956.0   
2              738.0  2068.0         2419.0         1623.0     1013.0  1456.0   
3              145.0   644.0          449.0         1331.0      185.0   872.0   
4              160.0   277.0          160.0          882.0       97.0   521.0   
...              ...     ...            ...            ...        ...     ...   
Some college     NaN     NaN            NaN            NaN        NaN     NaN   
White            NaN     NaN            NaN            NaN        NaN     NaN   
always           NaN     NaN            NaN            NaN        NaN     NaN   
rarely/never     NaN     NaN            NaN            NaN        NaN     NaN   
sporadic         NaN     NaN            NaN            NaN        NaN     NaN   

              Military Support  Respecting_Others  Believing in God  \
-1                        44.0               26.0              36.0   
1                       3684.0             3499.0            2758.0   
2                       1454.0             1938.0             931.0   
3                        459.0              297.0             810.0   
4                        195.0               76.0            1301.0   
...                        ...                ...               ...   
Some college               NaN                NaN               NaN   
White                      NaN                NaN               NaN   
always                     NaN                NaN               NaN   
rarely/never               NaN                NaN               NaN   
sporadic                   NaN                NaN               NaN   

              Protesting  ...  Racism_Policing  Society_Whole  \
-1                  38.0  ...             53.0           54.0   
1                 1854.0  ...           1708.0          990.0   
2                 2340.0  ...           1402.0         1618.0   
3                 1127.0  ...           1235.0         1519.0   
4                  477.0  ...           1438.0         1655.0   
...                  ...  ...              ...            ...   
Some college         NaN  ...              NaN            NaN   
White                NaN  ...              NaN            NaN   
always               NaN  ...              NaN            NaN   
rarely/never         NaN  ...              NaN            NaN   
sporadic             NaN  ...              NaN            NaN   

              Mainstream_Media  Way_People_Talk  ppage    educ    race  \
-1                        32.0             45.0    NaN     NaN     NaN   
1                       2605.0           1590.0    NaN     NaN     NaN   
2                       1800.0           2678.0    NaN     NaN     NaN   
3                        900.0           1229.0    NaN     NaN     NaN   
4                        499.0            294.0    NaN     NaN     NaN   
...                        ...              ...    ...     ...     ...   
Some college               NaN              NaN    NaN  1710.0     NaN   
White                      NaN              NaN    NaN     NaN  3710.0   
always                     NaN              NaN    NaN     NaN     NaN   
rarely/never               NaN              NaN    NaN     NaN     NaN   
sporadic                   NaN              NaN    NaN     NaN     NaN   

              gender  income_cat  voter_category  
-1               NaN         NaN             NaN  
1                NaN         NaN             NaN  
2                NaN         NaN             NaN  
3                NaN         NaN             NaN  
4                NaN         NaN             NaN  
...              ...         ...             ...  
Some college     NaN         NaN             NaN  
White            NaN         NaN             NaN  
always           NaN         NaN          1811.0  
rarely/n

# Perform exploratory analysis & visualization using Matplotlib and Seaborn

# Ask and Answer questions about the data in a Jupyter Notebook

- Do people of certain backgrounds tend to think differently American values?
- Do certain answers correlate with each other?

# Summarize your inference and write a conclusion


In [9]:
# 